sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Config/BlobsExtensionConfigProvider.cs

```diff
@@ -95,10 +92,10 @@ private void InitializeBlobTriggerBindings(ExtensionConfigContext context)
             context.AddConverter<BlobBaseClient, Stream>(ConvertToStreamAsync);
 
             // Blob type is a property of an existing blob.
-            // $$$ did we lose CloudBlob. That's a base class for Cloud*Blob, but does not implement ICloudBlob?
             context.AddConverter(new StorageBlobConverter<AppendBlobClient>());
             context.AddConverter(new StorageBlobConverter<BlockBlobClient>());
             context.AddConverter(new StorageBlobConverter<PageBlobClient>());
+            context.AddConverter<BlobBaseClient, BlobClient>(ConvertBlobBaseClientToBlobClient);
```

Comment:
When would this converter be invoked?

Reply:
There are two cases:
- When somebody binds to collection of blobs using `[Blob] IEnumerable<BlobClient>` [here](https://github.com/kasobol-msft/azure-sdk-for-net/blob/5ea0b2d1f8153ebdc54f45a8892228757df6fd26/sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Config/BlobsExtensionConfigProvider.cs#L215). In this case this will go through "identity conversion" (i.e. it will already resolve correct type few lines above).
- The second one is if someone uses `[BlobTrigger] BlobClient`. In that case the trigger may receive "invoke string" (which is basically blob path) or [this](https://github.com/Azure/azure-sdk-for-net/blob/1891cfca3fafed924da0eb76afecfb7cc11022ef/sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Listeners/BlobQueueTriggerExecutor.cs#L128) which is coming from [here](https://github.com/Azure/azure-sdk-for-net/blob/651bd6550709bde2ee28e81f451a556c542ccfd4/sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Bindings/StorageBlobContainerExtensions.cs#L40-L54).

Now that you ask this question I can see that in first case we could replace this converter with `new StorageBlobConverter<BlobClient>()` which basically type casts and validates that resolved type matches requested type. That'd be doable as we already know target type.
If we pursue that I think it would be good to move converter registrations from `context` to each `rule` so that `[Blob]` and `[BlobTrigger]` have their own sets of converters defined separately (even if it means small duplication).

As for the second case. We won't get rid of that converter as at the time the client is resolved we don't know the target type, so we can't make a decision whether to resolve BlobClient or BlockBlobClient. Hence conversion BlockBlobClient->BlobClient  or BlobClient ->BlockBlobClient (if we swap everything) is going to be somewhere.

Let me know what you think.

Reply:
I went ahead and made these changes. Live would be easier if we could traverse from XBlobClient to ContainerClient like we could in V11.